Rewrite long-range actors

jngrad opened the pull request:
Fixes #4454

Description of changes:
- apply the SOLID principles
   - split electrostatics from magnetostatics
   - factor out code duplication
   - extract P3M/DP3M tuning code
   - hide ELC details in private class methods (the energy/force correction logic used to be exposed in `coulomb.cpp`)
   - hide Coulomb implementation details by no longer indirectly including `coulomb.hpp` in 30 unrelated files in the core
- write clear pre-conditions as event hooks
   - these events are triggered upon box size change, periodicity change, node grid change and cell structure change
      - partial fix for #4428
      - setting a `node_grid` incompatible with a long-range actor is no longer possible (the old `node_grid` is restored)
   - these preconditions always run once before the integration loop starts
   - write test cases to check for these events
   - bugfix: charge neutrality checks now run once before the integration loop starts, instead of once at actor activation
- encapsulate the state of long-range methods in classes managed by the script interface
   - bugfix: the cached `DipolarP3M` energy correction is now invalidated when the box length changes
   - remove global variables for CPU methods (partial fix for #2628) and introduce one global variable for each type of long-range actor: dipolar solver, Coulomb solver, Coulomb extension
  - convert Cython files for long-range actors to Python files using the `ScriptInterface` framework
     - long-range actors and their python object now have the same lifetime, regardless of whether the actor is active or not (partial fix for #4220)
  - the list of active long-range actors can no longer end up in an invalid state (partial fix for #4219):
     - updating an active actor with an invalid parameter now automatically restores the original parameter
     - inserting an invalid long-range actor in the list of actors now automatically removes the actor from the list of active actors, even when the exception happens during tuning
- API changes:
   - `ELC` and `DLC` are no longer extensions, instead they take another actor as argument
   - `P3MGPU` is now able to execute the CPU energy and pressure kernels using parameters tuned for the GPU force kernel
      - closes #239
      - P3M-based reaction scripts run with 2x speed-up when reaction steps take place every 1000 integration steps
   - `P3MGPU` is now supported by `ELC`
   - long-range actor parameters are now immutable, since changing one parameter (e.g. the prefactor) would usually invalidate the tuned parameters
   - the charge neutrality check sensitivity can now be modified via `actor.charge_neutrality_tolerance`; this is mostly relevant to actors coupled to `ICC`, whose induced charges can have value spanning several orders of magnitude
- under-the-hood changes:
   - CPU-based long-range actors no longer rely on the `MpiCallbacks` framework
   - the subtracted Coulomb bonded interaction and pair criteria no longer indirectly include `coulomb.hpp`

reinaual commented on src/core/electrostatics/icc.cpp:
```diff
@@ -0,0 +1,283 @@
+    return;
+  }
+
+  auto prefactor = 0.;
+  if (electrostatics_actor) {
+    prefactor =
+        boost::apply_visitor(GetCoulombPrefactor{}, *electrostatics_actor);
+  }
+  assert(prefactor != 0.);
```
Just out of interest, is there a way to reach that point with no `electrostatics_actor` activated? In my ideal world ICC*-iterations should never be executed without an electrostatics solver. That would make the `if (electrostatics_actor)` to always evalute to `true`, right?

reinaual commented on src/core/unit_tests/EspressoSystemStandAlone_test.cpp:
```diff
@@ -245,12 +265,8 @@ BOOST_FIXTURE_TEST_CASE(espresso_system_stand_alone, ParticleFactory,
-    p3m_set_params(3.5, mesh.data(), 5, 0.654, 1e-3);
-    p3m_set_mesh_offset(0.5, 0.5, 0.5);
+    mpi_set_tuned_p3m();
+    auto const prefactor = 2.;
```
I'm not to sure if it is a good idea to rely on this variable to be the same as the hard-coded one above in
`auto solver = std::make_shared<CoulombP3M>(std::move(p3m), 2., 1, false);`

jngrad replied:
good point, this is fixed in the latest commit

reinaual commented on src/python/espressomd/actors.py:
```diff
@@ -0,0 +1,99 @@
+        if actor in Actors.active_actors:
+            raise highlander.ThereCanOnlyBeOne(actor)
+
+        if isinstance(actor, script_interface.ScriptInterfaceHelper):
+            actor._activate()
+
+        self.active_actors.append(actor)
+
+        if not isinstance(actor, script_interface.ScriptInterfaceHelper):
+            actor._activate()
```
where does it come from, that for `ScriptInterfaceHelper`-objects we need to call `_activate()` before appending it to the list?

jngrad replied:
Here is the intended workflow:
- `_activate()` calls the C++ actor registration code that adds the actor in the global optional variant
- the now active actor in the core runs the sanity checks and tuning algorithm (if the method supports tuning)
- if checks or tuning fail, a C++ exception is thrown, which is handled in the [registration code](https://github.com/espressomd/espresso/pull/4506/files/0d6f2c1ba994eeac24eb41c0b11581d668811b07#diff-f9f14648861a7b0f5fa3655eedb94ffb693693d1e0410be86582aae1cdb47d38R54-R62)
   - the actor is safely removed from the core global
   - the exception is re-thrown
   - on MPI rank 0, the C++ exception becomes a Python exception
   - line 54 in actors.py throws
   - the python object is never added to the list of active actors

Since removing an actor from the actor list will trigger sanity checks in the core, we cannot remove a python actor if it is not also in the core. We also cannot keep an invalid actor in the core. LB and EK actors still rely on the old workflow.

reinaual commented on src/python/espressomd/lb.pyx:
```diff
@@ -24,19 +24,160 @@ import functools
+        if inter in FluidActor.active_list:
+            if not FluidActor.active_list[inter]:
+                raise Exception(
+                    "Class not registered in Actor.active_list: " + self.__class__.__bases__[0].__name__)
```
What is the reason we use `+` and `%`-operators for strings instead of f-strings in this whole file?

jngrad replied:
I had moved the original actor code without modifying it. It's now using f-strings everywhere in the latest commit.

reinaual commented on src/script_interface/electrostatics/ICCStar.hpp:
```diff
@@ -0,0 +1,122 @@
+
+public:
+  ICCStar() {
+    add_parameters({
+        {"n_icc", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.n_icc; }},
+        {"max_iterations", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.max_iterations; }},
+        {"eps_out", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.eps_out; }},
+        {"areas", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.areas; }},
+        {"epsilons", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.epsilons; }},
+        {"sigmas", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.sigmas; }},
+        {"convergence", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.convergence; }},
+        {"normals", AutoParameter::read_only,
+         [this]() {
+           return make_vector_of_variants(actor()->icc_cfg.normals);
+         }},
+        {"ext_field", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.ext_field; }},
+        {"relaxation", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.relaxation; }},
+        {"citeration", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.citeration; }},
+        {"first_id", AutoParameter::read_only,
+         [this]() { return actor()->icc_cfg.first_id; }},
+    });
```
Technically it would not be necessary to prevent users from overwriting any parameter here (except for `citeration`). Is there a reason we don't allow it?

jngrad replied:
Actually, most of these parameters depend on one another. For example, we cannot append a value to `sigmas` because it would have a different length as `epsilons` and `areas`. We also cannot move the vector length checks from the constructor to a new method `check_vector_consistency`, because from the python interface we can only update ICC properties one by one. We would need to create a lock system, like a context manager, to guarantee that a set of changes is self-consistent, and if it is not roll back to the old properties. That's doable, but then what should we do if checkpointing is triggered by a `Ctrl+C` signal in the middle of the context manager?

I chose this solution because it seems easier to me to create a new ICC object with the new properties to replace the existing one. Let me know if there is a use case for dynamically changing these properties during the simulation.

reinaual replied:
okay, i see the trouble with that... altough the only criterion that has to be fulfilled is that `sigmas, areas, epsilons, normals` have to have a length `>= n_icc` for ICC to work properly. It only uses the first `n_icc` values of the arrays anyway, so they could also each have a different length, as long as they are long enough.

AFAIK there is no real use-case right now for dynamically changing the amount of ICC charges, although I did have a look at such systems once where i did ICC-refinements. What i would argue should be changeable are all the remaining parameters that don't have any side-effect outside of ICC and can't leave it in a problematic state. This includes `max_iterations, eps_out, convergence, ext_field, relaxation` (personally I don't mind about `first_id`, altough this also has no side-effect outside of ICC). For the `ext_field` parameter I could also see a use-case, e.g. for alternating electric fields. Probably this would also be close to no effort to implement.

TL;DR: How about we leave all ICC particle-data immutable, but one can change all the other stuff that is independent from each other and has no influence outside of ICC?

reinaual commented on src/core/tuning.cpp:
```diff
@@ -39,36 +37,63 @@
-  return 1000. * running_average.avg();
+  /* MPI returns in seconds, returned value should be in ms. */
+  auto retval = 1000. * running_average.avg();
+  boost::mpi::broadcast(comm_cart, retval, 0);
+  return retval;
```
Am i right that it wasn't necessary to broadcast this value before because the timing only ran on the head-node?

jngrad replied:
that's correct; now all MPI ranks must have the same timings, otherwise the bisection will diverge

jngrad commented on src/scafacos/src/Coulomb.cpp:
```diff
@@ -0,0 +1,97 @@
+  fields.resize(3ul * n_part);
+  potentials.resize(n_part);
+
+  tune(charges, positions);
+  auto const size = static_cast<int>(n_part);
+  handle_error(fcs_run(m_handle, size, positions.data(), charges.data(),
+                       fields.data(), potentials.data()));
```
Does anyone remember why in the original implementation, we need to re-tune ScaFaCoS before each integration step? Maybe it has to do with side effects when the `node_grid` or number of particles changes, or when we reload from a checkpoint (tuned parameters cannot be checkpointed). Commenting out line 84 doesn't affect the testsuite, but the edge case it's supposedly guarding against is not documented in the original implementation.

reinaual reviewed (commented):
LGTM

jngrad committed:
Rewrite long-range actors

Create more fine-grained events for long-range actors to react to
changes in the cell structure and box geometry. Convert Cython code
to Python code with the ScriptInterface framework and remove global
variables of CPU methods. Fix regressions in the energy and pressure
analysis module. Split electrostatic and magnetostatic methods.
